src/main/cpp/native_glue.cc

```diff
@@ -19,140 +19,72 @@
 #include "native_glue.hh"
 #include <queue>
 #include <new>
+#include <codecvt>
+#include <locale>
 
 using namespace litecore;
 using namespace litecore::jni;
 using namespace std;
 
-namespace litecore { namespace jni {
-        void UTF8CharToModifiedUTF8(const char *input, char *output);
-        void ModifiedUTF8ToUTF8(char* input);
-        void ModifiedUTF8CharToUTF8(char* input);
+namespace litecore {
+    namespace jni {
+        std::string JstringToUTF8(JNIEnv *env, jstring jstr);
+        jstring UTF8ToJstring(JNIEnv *env, char *s, size_t size);
     }
 }
 
-void litecore::jni::UTF8CharToModifiedUTF8(const char *input, char *output) {
-    char c = input[0];
-    char c1 = input[1] & 0x3F;
-    char c2 = input[2] & 0x3F;
-    char c3 = input[3] & 0x3F;
-
-    int unicodePoint = ((c & 0x07) << 18) | (c1 << 12) | (c2 << 6) | c3;
-    unicodePoint -= 0x10000;
-
-    int surrogates[2] { 0xD800 | (unicodePoint >> 10), 0xDC00 | (unicodePoint & 0x3FF) };
-
-    output[0] = (surrogates[0]>>12 & 0x0F) | 0xE0;
-    output[1] = (surrogates[0]>>6  & 0x3F) | 0x80;
-    output[2] = (surrogates[0]     & 0x3F) | 0x80;
-    output[3] = (surrogates[1]>>12 & 0x0F) | 0xE0;
-    output[4] = (surrogates[1]>>6  & 0x3F) | 0x80;
-    output[5] = (surrogates[1]     & 0x3F) | 0x80;
-}
-
-ssize_t litecore::jni::UTF8ToModifiedUTF8(const char* input, const char** output, size_t len){
-    // https://github.com/android-ndk/ndk/issues/283
-    size_t extraBytes = 0;
-    const auto unsignedInput = (const uint8_t *)input;
-
-    // Need to figure out the actual length since each
-    // 4-bytes of real UTF-8 is 6 bytes of modified UTF-8
-    // but convert the bytes while we are at it
-    for(size_t i = 0; i < len; i++) {
-        if(unsignedInput[i] >= 0xF0) {
-            // 0xF0 and above marks 4-byte sequences
-            extraBytes += 2;
-
-            // 3 + 1 from next loop iteration = 4 bytes advance
-            i += 3;
-        }
-    }
-
-    if(extraBytes == 0) {
-        // No modifications necessary
-        *output = nullptr;
-        return len;
-    }
-
-    size_t newStrLen = len + extraBytes;
-    char* newBytes = (char *)malloc(newStrLen);
-    if(newBytes == nullptr) {
-        *output = nullptr;
-        return -1;
-    }
-
-    int offset = 0;
-    for(size_t i = 0; i < len; i++) {
-        if(unsignedInput[i] >= 0xF0) {
-            UTF8CharToModifiedUTF8(input + i, newBytes + i + offset);
-            i += 3;
-            offset += 2;
-        } else {
-            newBytes[i+offset] = unsignedInput[i];
-        }
+// Java uses Modified-UTF-8, not UTF-8: Attempting to decode a real UTF-8 string will cause a failure that looks like:
```

Comment:
Is this true of all JVMs, not just Android's? If not, should there be a [compile-time?] flag to skip expensive encoding conversions when not needed?

Reply:
I am not clear on this.  There is just very little good information about the issue.  I originally saw things that gave me the impression that the problem only existed on certain versions of Android.  Since then, I've seen things that suggest that the problem applies to Java in general.

I might have to try test cases...

Reply:
I verified, using only documentation, that Modified UTF-8 is, in fact, a Java standard, not an Android standard.  I have not verified that it applies to all implementations of the JNI but I believe that always using this code is justified.